### DISCUSSION TITLE
Pico W: `OSError: -1` when using SSL (`MQTTClient(..., ssl=True)`)

### DISCUSSION THREAD

**Discussion by sgbaird** [micropython/micropython#10182] [Libraries & Drivers]
The following works without issue (for context, see https://github.com/sparks-baird/self-driving-lab-demo/blob/main/src/public_mqtt_sdl_demo/main.py):
```python
...
host = "test.mosquitto.org"

client = MQTTClient(
    prefix,
    host,
    user=None,
    password=None,
    keepalive=0,
    ssl=False,
    ssl_params={},
)
...
while True:
    client.check_msg()
    ...
```

Setting `ssl=True` per the comment in https://github.com/micropython/micropython/discussions/9259#discussioncomment-3831835 causes an error after a short period of time (less than a minute I think, I verified that it's **not** during the first call to `wait_msg`):
```python
...
host = "test.mosquitto.org"

client = MQTTClient(
    prefix,
    host,
    user=None,
    password=None,
    keepalive=0,
    ssl=True,
    ssl_params={"server_hostname": host}, # commenting this line out, but leaving ssl=True results in same error
)
...
while True:
    client.check_msg()
    ...
```

```python
Traceback (most recent call last):
  File "<stdin>", line 328, in <module>
  File "/lib/umqtt/simple.py", line 214, in check_msg
  File "/lib/umqtt/simple.py", line 183, in wait_msg
OSError: -1
```

Here is how `wait_msg` is defined (note added on line 183):
```python
    # Wait for a single incoming MQTT message and process it.
    # Subscribed messages are delivered to a callback previously
    # set by .set_callback() method. Other (internal) MQTT
    # messages processed internally.
    def wait_msg(self):
        res = self.sock.read(1)
        self.sock.setblocking(True)
        if res is None:
            return None
        if res == b"":
            raise OSError(-1) ##### LINE 183 #####
        if res == b"\xd0":  # PINGRESP
            sz = self.sock.read(1)[0]
            assert sz == 0
            return None
        op = res[0]
        if op & 0xF0 != 0x30:
            return op
        sz = self._recv_len()
        topic_len = self.sock.read(2)
        topic_len = (topic_len[0] << 8) | topic_len[1]
        topic = self.sock.read(topic_len)
        sz -= topic_len + 2
        if op & 6:
            pid = self.sock.read(2)
            pid = pid[0] << 8 | pid[1]
            sz -= 2
        msg = self.sock.read(sz)
        self.cb(topic, msg)
        if op & 6 == 2:
            pkt = bytearray(b"\x40\x02\0\0")
            struct.pack_into("!H", pkt, 2, pid)
            self.sock.write(pkt)
        elif op & 6 == 4:
            assert 0
```

I also tried adding `sleep(0.1)` in the while loop, but got the same error.

Related:
- https://github.com/micropython/micropython/discussions/9222 by @IanW6374
- https://github.com/micropython/micropython/discussions/9259 by @IanW6374

@Carglglz

**Comment by Carglglz** [comment]
@sgbaird 
I was able to reproduce this with an esp32, I think is due to the mqtt broker closing the connection on its side, look for keepalive parameter and doing pings in while loop e.g.:
```py
....
keepalive=30
....
while True:
      client.check_msg()
      client.ping()
      time.sleep(0.5)
```
This should fix it 👀 ...

**Reply by sgbaird** [reply]
I had a ["heartbeat" (ping) function](https://github.com/sparks-baird/self-driving-lab-demo/blob/abca9bcdd32f34c2b2b6a4991adb215942d97499/src/public_mqtt_sdl_demo/main.py#L255-L263) (every 300 s) and `keepalive=0`. Per your suggestion, after reducing the ping interval to `15` seconds and setting `keepalive=30`, it seems to be running normally. Thank you!
